miden-tx/src/host/tx_authenticator.rs

```diff
@@ -0,0 +1,158 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+
+use miden_objects::{
+    accounts::AccountDelta,
+    crypto::dsa::rpo_falcon512::{self, Polynomial},
+};
+use rand::Rng;
+use vm_processor::{Digest, Felt, Word};
+
+use crate::error::AuthenticationError;
+
+// TRANSACTION AUTHENTICATOR
+// ================================================================================================
+
+/// Represents an authenticator for transactions.
+///
+/// Its main use is to provide a method to create a DSA signature for a message,
+/// based on an [AccountDelta].
+///
+/// The signature is intended to provide authentication. The implementer can verify
+/// and approve any changes before the message is signed.
+pub trait TransactionAuthenticator {
+    /// Retrieves a signataure for a specific message as a list of [Felt].
+    /// The request is initiaed by the VM as a consequence of the SigToStack advice
+    /// injector.
+    ///
+    /// - `pub_key`: The public key used for signature generation.
+    /// - `message`: The message to sign, usually a commitment to the transaction data.
+    /// - `account_delta`: An informational parameter describing the changes made to
+    ///   the account up to the point of calling `get_signature()`. This allows the
+    ///   authenticator to review any alterations to the account prior to signing.
+    ///   It should not be directly used in the signature computation.
+    fn get_signature(
+        &mut self,
+        pub_key: Word,
+        message: Word,
+        account_delta: &AccountDelta,
+    ) -> Result<Vec<Felt>, AuthenticationError>;
```

Comment:
I wonder if it is possible to not pass `self` as mutable here and handle potential mutability requirements via interior mutability within the specific implementation. This would simplify the external interface and would prevent mutability requirements from (as you noted) bubbling up further up the chain. But also, not all signature schemes require this (e.g., if someone were to implement the authenticator only for deterministic signature schemes, they won't need the mutable reference to `self`). So, it feels like mutable `self` is a requirement of a specific implementation rather than the overall scheme.

Reply:
Yeah, I think this is better after seeing the consequences of making it a mutable reference. Done.